### ISSUE TITLE
Navigation block: decouple ESC key JavaScript action from focus trap

### ISSUE THREAD

**Issue by webmandesign** [WordPress/gutenberg#63033]
### Description

First, I'm not sure whether this is a bug or a feature request.

During [accessibility review of my theme](https://themes.trac.wordpress.org/ticket/173979#comment:21) there are interesting points regarding the focus trap issue in mobile menu (the overlay menu of Navigation block). (Check the ["Finding 2" and "Finding 3"](https://themes.trac.wordpress.org/ticket/173979#comment:21) for more info.)

Unlike **Twenty Twenty-Four** theme, my **Zooey** theme displays mobile submenus (the submenus in overlay menu mode of Navigation block) _collapsed_. I decided to do so as the functionality is still enabled in the Navigation block, and it is just a matter of CSS styles to re-apply it for overlay menus.

However, this produces an accessibility issue of focus trap when submenu is expanded in overlay menu.

From my debugging it seems to be triggered by [JavaScript `handleMenuKeydown()` action](https://github.com/WordPress/gutenberg/blob/v18.6.1/packages/block-library/src/navigation/view.js#L109) applied both on overlay menu wrapper and [on all submenu containers](https://github.com/WordPress/gutenberg/blob/v18.6.1/packages/block-library/src/navigation/index.php#L830).

If I understand correctly, the `handleMenuKeydown()` action is applied on submenu containers only for enabling ESC key action on **non**-overlay menus. It was never intended to create a focus trap on submenus when overlay menu is opened.

Both ESC key action and focus trap are meant only for overlay/mobile menu container.

To prevent similar issues and allow theme authors to decide the submenu actions themselves if needed, please decouple the ESC key action from focus trap. Now they are both triggered by single [JavaScript `handleMenuKeydown()` action](https://github.com/WordPress/gutenberg/blob/v18.6.1/packages/block-library/src/navigation/view.js#L109), which makes it impossible for me to control the submenu behavior so it can pass the accessibility review.

I suggest decoupling the JavaScript action into 2 separate actions of `handleEscapeKey()` and `handleFocusTrap()`, and trigger these actions on elements with `data-wp-on--keydown-esc` and `data-wp-on--keydown-tab` attributes, respectively.

The current faulty behavior (based on description of the [accessibility review](https://themes.trac.wordpress.org/ticket/173979#comment:21)) can be previewed at [theme's PlayGround instance](https://playground.wordpress.net/#{%22preferredVersions%22:{%22php%22:%227.4%22,%22wp%22:%22latest%22},%22steps%22:[{%22step%22:%22login%22,%22username%22:%22admin%22,%22password%22:%22password%22},{%22step%22:%22defineWpConfigConsts%22,%22consts%22:{%22WP_DEBUG%22:true}},{%22step%22:%22importFile%22,%22file%22:{%22resource%22:%22url%22,%22url%22:%22https:\/\/raw.githubusercontent.com\/WordPress\/theme-test-data\/master\/themeunittestdata.wordpress.xml%22,%22caption%22:%22Downloading%20theme%20testing%20content%22},%22progress%22:{%22caption%22:%22Installing%20theme%20testing%20content%22}},{%22step%22:%22installPlugin%22,%22pluginZipFile%22:{%22resource%22:%22wordpress.org\/plugins%22,%22slug%22:%22theme-check%22},%22options%22:{%22activate%22:true}},{%22step%22:%22installTheme%22,%22themeZipFile%22:{%22resource%22:%22url%22,%22url%22:%22https:\/\/downloads.wordpress.org\/theme\/zooey.1.1.2.zip?nostats=1%22,%22caption%22:%22Downloading%20the%20theme%22}}]}).

It would be probably a good idea to provide theme authors [more information about Navigation block accessibility](https://developer.wordpress.org/themes/functionality/block-theme-accessibility/#accessible-navigation-menu) actions as debugging this issue was a bit tedious task for me.

Thank you for consideration!

### Step-by-step reproduction instructions

Based on description of the [accessibility review](https://themes.trac.wordpress.org/ticket/173979#comment:21), please check the behavior of expanded submenu focus trap in opened overlay menu of Navigation block in [theme's PlayGround instance](https://playground.wordpress.net/#{%22preferredVersions%22:{%22php%22:%227.4%22,%22wp%22:%22latest%22},%22steps%22:[{%22step%22:%22login%22,%22username%22:%22admin%22,%22password%22:%22password%22},{%22step%22:%22defineWpConfigConsts%22,%22consts%22:{%22WP_DEBUG%22:true}},{%22step%22:%22importFile%22,%22file%22:{%22resource%22:%22url%22,%22url%22:%22https:\/\/raw.githubusercontent.com\/WordPress\/theme-test-data\/master\/themeunittestdata.wordpress.xml%22,%22caption%22:%22Downloading%20theme%20testing%20content%22},%22progress%22:{%22caption%22:%22Installing%20theme%20testing%20content%22}},{%22step%22:%22installPlugin%22,%22pluginZipFile%22:{%22resource%22:%22wordpress.org\/plugins%22,%22slug%22:%22theme-check%22},%22options%22:{%22activate%22:true}},{%22step%22:%22installTheme%22,%22themeZipFile%22:{%22resource%22:%22url%22,%22url%22:%22https:\/\/downloads.wordpress.org\/theme\/zooey.1.1.2.zip?nostats=1%22,%22caption%22:%22Downloading%20the%20theme%22}}]}).

Use keyboard to open and close overlay/mobile menu and its submenus.

Once submenu is expanded, test the `SHIFT + TAB` key to see it does not work as focus trap JavaScript action recalculates first and last focusable element and applies focus trap incorrectly.

### Screenshots, screen recording, code snippet

You can check videos provided at [theme accessibility review comment](https://themes.trac.wordpress.org/ticket/173979#comment:21).

### Environment info

- WordPress 6.5.5
- Zooey theme 1.1.2
- Gutenberg 18.6.1

### Please confirm that you have searched existing issues in the repo.

Yes

### Please confirm that you have tested with all plugins deactivated except Gutenberg.

Yes

**Comment by webmandesign** [comment]
Hi @skorasaurus,
Do you know if anybody looked into this issue? It's getting old, but the issue still persists. Thank you!

**Comment by amberhinds** [comment]
@joedolson @afercia, do you know if this is still an issue that needs to be addressed?

**Comment by joedolson** [comment]
It looks like there's a duplicate issue raised #78355, and that issue has a PR [78379](https://github.com/WordPress/gutenberg/pull/78379)

But nothing is merged, and the PR could probably stand updating & testing.

**Comment by joedolson** [comment]
I didn't explore in enough detail to know for sure that's an expect duplicate, but it's definitely related.

**Comment by webmandesign** [comment]
Hi @amberhinds, @joedolson,

The #78355 doesn't seem to be a duplicate of this issue. Although, that issue is also interesting and I've personally had a theme user recently report it on my support forum. Glad to see this will be fixed in core.

Back to my original issue here, I've retested the issue with current WP7.0 and Zooey 2.0.1. Here are my findings and steps on how to reproduce them:


### Prerequisites

1. Use [WP Playground](https://playground.wordpress.net/#%7B%22preferredVersions%22%3A%7B%22php%22%3A%228.3%22%2C%22wp%22%3A%22latest%22%7D%2C%22steps%22%3A%5B%7B%22step%22%3A%22login%22%2C%22username%22%3A%22admin%22%2C%22password%22%3A%22password%22%7D%2C%7B%22step%22%3A%22defineWpConfigConsts%22%2C%22consts%22%3A%7B%22WP_DEBUG%22%3Atrue%7D%7D%2C%7B%22step%22%3A%22importFile%22%2C%22file%22%3A%7B%22resource%22%3A%22url%22%2C%22url%22%3A%22https%3A%2F%2Fraw.githubusercontent.com%2FWordPress%2Ftheme-test-data%2Fmaster%2Fthemeunittestdata.wordpress.xml%22%2C%22caption%22%3A%22Downloading%20theme%20testing%20content%22%7D%2C%22progress%22%3A%7B%22caption%22%3A%22Installing%20theme%20testing%20content%22%7D%7D%2C%7B%22step%22%3A%22installTheme%22%2C%22themeZipFile%22%3A%7B%22resource%22%3A%22url%22%2C%22url%22%3A%22https%3A%2F%2Fdownloads.wordpress.org%2Ftheme%2Fzooey.2.0.1.zip%3Fnostats%3D1%22%2C%22caption%22%3A%22Downloading%20the%20theme%22%7D%7D%5D%7D) with Zooey theme and theme unit test data installed
2. Navigate to **Appearance → Customize → Theme Options → Accessibility** and toggle off the **"Expand mobile navigation submenus"** option, and save customizer changes
3. Now go to the website front-end and interact with hamburger menu as per following instructions below. (Also, make sure the hamburger menu displays a pagelist.)

---

### 1. ISSUE: Focus trap not always working in submenu ([Finding 2](https://themes.trac.wordpress.org/ticket/173979#comment:21))

Original report:
> When the last element in a submenu is another submenu, the focus trap inside the submenu will not work and just jump to the next main menu item.
> Screen capture: https://themes.trac.wordpress.org/attachment/ticket/173979/173979_13062024_01.mov

I can no longer reproduce this issue, so it seems to be fixed from my point of view.

---

### 2. ISSUE: Backwards tabbing broken after closing submenu ([Finding 3](https://themes.trac.wordpress.org/ticket/173979#comment:21))

1. Open overlay menu
2. Tab to "Level 1" menu item, tab once more to focus the "+" submenu toggle button
3. Press spacebar to open the submenu (focus stays on "-" submenu toggle button)
4. Press spacebar again to close the submenu (focus stays on "+" submenu toggle button)
5. Press SHIFT+TAB so the focus moves to "Level 1" menu item
6. Press SHIFT+TAB again:
	- _expected_: focus should move to previous "About The Tests" menu item
	- **error**: focus stays on "Level 1" menu item no matter how many times one press SHIFT+TAB

---

### 3. ISSUE: Focus trap within overlay navigation brakes after closing submenu

1. Open overlay menu
2. Tab to "Level 1" menu item, tab once more to focus the "+" submenu toggle button
3. Press spacebar to open the submenu (focus stays on "-" submenu toggle button)
4. Press spacebar again to close the submenu (focus stays on "+" submenu toggle button)
5. Now press TAB multiple times, beyond the last (top-level) menu item:
	- _expected_: focus trap should move the focus back to "⨉" overlay menu toggle button after the last menu item
	- **error**: focus goes outside the overlay menu, bypassing the focus trap

---

### 4. POSSIBLE ISSUE: Focus trap not staying within submenu after closing sub-submenu

I'm not sure whether this is an issue or not. I presume once the submenu is open, the focus trap should stay within the submenu. Please, correct me if I'm wrong, and in that case disregard this specific report.

1. Open overlay menu
2. Tab to "Level 1" menu item, tab once more to focus the "+" submenu toggle button
3. Press spacebar to open the submenu (focus stays on "-" submenu toggle button)
4. Now press TAB multiple times to see that the focus is _correctly_ trapped within submenu
5. Tab to "Level 2" menu item, tab once more to focus the "+" sub-submenu toggle button and toggle it pressing spacebar
6. Now press TAB multiple times to see that the focus is _correctly_ trapped within sub-submenu
7. Tab to sub-submenu toggle button and close it by pressing spacebar
8. Now press TAB multiple times:
	- _expected_: focus should be trapped within (parent) submenu
	- **error**: focus goes outside the (parent) submenu to subsequent "Lorem ipsum" top-level menu item


---

These issues are present even when using WP7.0+ Navigation Overlay.

**Comment by amberhinds** [comment]
Thanks for the additional information @webmandesign!
